x/config/configurator.go

```diff
@@ -0,0 +1,648 @@
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if spec.BuildOnewayOutbound != nil {
+		out.onewayOutboundBuilder, err = newOutboundBuilder(spec.BuildOnewayOutbound, _typeOfOnewayOutbound)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	out.unaryOutboundPresets, err = buildOutboundPresets(spec.UnaryOutboundPresets, _typeOfUnaryOutbound)
+	if err != nil {
+		return nil, err
+	}
+
+	out.onewayOutboundPresets, err = buildOutboundPresets(spec.OnewayOutboundPresets, _typeOfOnewayOutbound)
+	if err != nil {
+		return nil, err
+	}
+
+	return &out, nil
+}
+
+func (s *transportSpec) TransportBuilder(attrs attributeMap) (TransportBuilder, error) {
+	return s.transportBuilder.Load(attrs)
+}
+
+func (s *transportSpec) InboundBuilder(attrs attributeMap) (InboundBuilder, error) {
+	if s.inboundBuilder == nil {
+		return nil, fmt.Errorf("transport %q does not define an inbound", s.Name)
+	}
+	return s.inboundBuilder.Load(attrs)
+}
+
+func (s *transportSpec) UnaryOutboundBuilder(preset string, attrs attributeMap) (UnaryOutboundBuilder, error) {
+	if s.unaryOutboundBuilder == nil {
+		return nil, fmt.Errorf("transport %q does not define a unary outbound", s.Name)
+	}
+
+	b := s.unaryOutboundBuilder
+	if preset != "" {
+		var ok bool
+		b, ok = s.unaryOutboundPresets[preset]
+		if !ok {
+			return nil, fmt.Errorf("unknown preset %q for unary outbound %q", preset, s.Name)
+		}
+	}
+
+	b, err := b.Load(attrs)
+	return b, err
+}
+
+func (s *transportSpec) OnewayOutboundBuilder(preset string, attrs attributeMap) (OnewayOutboundBuilder, error) {
+	if s.onewayOutboundBuilder == nil {
+		return nil, fmt.Errorf("transport %q does not define a oneway outbound", s.Name)
+	}
+
+	b := s.onewayOutboundBuilder
+	if preset != "" {
+		var ok bool
+		b, ok = s.onewayOutboundPresets[preset]
+		if !ok {
+			return nil, fmt.Errorf("unknown preset %q for oneway outbound %q", preset, s.Name)
+		}
+	}
+
+	b, err := b.Load(attrs)
+	return b, err
+}
+
+func (s *transportSpec) SupportsUnaryOutbound() bool {
+	return s.unaryOutboundBuilder != nil
+}
+
+func (s *transportSpec) SupportsOnewayOutbound() bool {
+	return s.onewayOutboundBuilder != nil
+}
+
+func buildOutboundPresets(m map[string]interface{}, resultType reflect.Type) (map[string]*builder, error) {
+	if len(m) == 0 {
+		return nil, nil
+	}
+	out := make(map[string]*builder, len(m))
+	for k, v := range m {
+		var err error
+		out[k], err = newOutboundBuilder(v, resultType)
+		if err != nil {
+			return nil, fmt.Errorf("invalid preset %q: %v", k, err)
+		}
+	}
+	return out, nil
+}
+
+// Returns a list of struct fields for the given type. The type may be a
+// struct or a pointer to a struct (arbitrarily deep).
+func fieldNames(t reflect.Type) map[string]struct{} {
+	for ; t.Kind() == reflect.Ptr; t = t.Elem() {
+	}
+
+	if t.Kind() != reflect.Struct {
+		return nil
+	}
+
+	fields := make(map[string]struct{}, t.NumField())
+	for i := 0; i < t.NumField(); i++ {
+		fields[t.Field(i).Name] = struct{}{}
```

Comment:
Good catch. This isn't actually used for anything but to verify that they don't have field names that collide with our reserved names (`With`, `Disabled`, etc.) but changing this now makes it more future proof.